include/seastar/rpc/rpc.hh

```diff
@@ -96,7 +97,43 @@ struct resource_limits {
     size_t max_memory = rpc_semaphore::max_counter(); ///< Maximum amount of memory that may be consumed by all requests
     /// Configures isolation for a connection based on its isolation cookie. May throw,
     /// in which case the connection will be terminated.
-    std::function<isolation_config (sstring isolation_cookie)> isolate_connection = default_isolate_connection;
+    class isolate_connection_wrapper {
+        using syncronous_isolation_function = std::function<isolation_config (sstring isolation_cookie)>;
+        using asyncronous_isolation_function = std::function<future<isolation_config> (sstring isolation_cookie)>;
+        using isolation_function_alternatives = std::variant<syncronous_isolation_function, asyncronous_isolation_function>;
+        isolation_function_alternatives _func;
+        struct isolation_function_visitor {
+                  isolation_function_visitor(const sstring& isolation_cookie) :
+                        _isolation_cookie(isolation_cookie) { }
+                  future<isolation_config> operator() (syncronous_isolation_function f) const {
+                      return futurize_invoke(f, _isolation_cookie);
+                  }
+                  future<isolation_config> operator() (asyncronous_isolation_function f) const {
+                      return f(_isolation_cookie);
+                  }
+              private:
+                  sstring _isolation_cookie;
+              };
+    public:
+        isolate_connection_wrapper() = default;
+        isolate_connection_wrapper(const syncronous_isolation_function& func) : _func{func} {
+        }
+        isolate_connection_wrapper(const asyncronous_isolation_function& func) : _func{func} {
+        }
+        template<typename T>
```

Comment:
Is isolate_connection_wrapper truly needed? Lots of code to wrap a variant. Can't we use the variant directly?

Reply:
We can use directly, as it was before. @denesb asked for a wrapper.

Reply:
I asked for a wrapper with the hope that we can do away with the variant, by automatically wrapping the synhronous version in a wrapper lambda, such that its signature will be identical to the async version, hence we can get away with a single type. The current wrapper as-is indeed doesn't add much. Sorry for the confusion.

Reply:
Ahhhh, I tried that and it seams to be pretty complicate (if you would like to keep the assignment untouched for backward compatibility). I guess it is doable with some templating and deduction trickery. I might work on something to be put in
future utils as a follow up.